File: src/main/java/com/noticemc/noticeconnect/files/ObjectMapperExample.java
/*
 *  <NoticeConnect>-<A login message plugin that runs on Velocity>
 *
 *  Written in 2021  by Nikomaru <nikomaru@nikomaru.dev>
 *
 *      To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the
 * public domain worldwide.
 *      This software is distributed without any warranty.
 *
 *      You should have received a copy of the CC0 Public Domain Dedication along with this software.
 *      If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */

package com.noticemc.noticeconnect.files;

import org.spongepowered.configurate.CommentedConfigurationNode;
import org.spongepowered.configurate.ConfigurateException;
import org.spongepowered.configurate.hocon.HoconConfigurationLoader;
import org.spongepowered.configurate.objectmapping.ConfigSerializable;
import org.spongepowered.configurate.objectmapping.meta.Comment;

import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

public final class ObjectMapperExample {

    public static void main(Path path) throws ConfigurateException {
        final HoconConfigurationLoader loader = HoconConfigurationLoader.builder()
                .path(path) // or url(), or source/sink
                .build();

        final CommentedConfigurationNode node = loader.load(); // Load from file
        final Database database = new Database(); // Populate object
        final Message message = new Message();
        final Replace replace = new Replace();
        final Discord discord = new Discord();
        final DiscordMessage discordMessage = new DiscordMessage();

        // Do whatever actions with the configuration, then...
        node.node("message").set(message);
        node.node("database").set(database); // Update the backing node
        node.set(replace);
        node.node("discord").set(discord).node("message").set(discordMessage);
        // the backing node
        loader.save(node); // Write to the original file
    }


    @ConfigSerializable
    static class Database {

        private final String database = "NoticeConnect";

    }

    @ConfigSerializable
    static class Message {

        private final String firstJoin = "<gray>[<yellow>初見さんいらっしゃい</yellow>] <light_purple><name>さんがはじめて<currentServerName>サーバーに入りました!";

        private final String join = "<yellow><gold><name></gold>さんが<gold>サーバー<gold><gray>(<currentServerName>)</gray>にやってきました!";

        private final String left = "<yellow><gold><name></gold>さんが<gold>サーバー<gold><gray>(<currentServerName>)</gray>から離れました";

    }

    @ConfigSerializable
    static class Replace {

        @Comment("Example: test = \"テスト\" It needs to be written like this")
        private final Map<String, String> replace = new HashMap<>();


    }

    @ConfigSerializable
    static class Discord {

        private final String token = "";

        private final String channelId = "";

    }


    @ConfigSerializable
    static class DiscordMessage {

        private final String firstJoin = "%(PlayerName) さんがサーバーに初参加です";

        private final String join = "%(PlayerName) さんがサーバーに参加しました";

        private final String left = "%(PlayerName) さんがサーバーから退出しました";

    }
}

File: src/main/java/com/noticemc/noticeconnect/files/ObjectMapperExample.kt
/*
 *  <NoticeConnect>-<A login message plugin that runs on Velocity>
 *
 *  Written in 2021  by Nikomaru <nikomaru@nikomaru.dev>
 *
 *      To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the
 * public domain worldwide.
 *      This software is distributed without any warranty.
 *
 *      You should have received a copy of the CC0 Public Domain Dedication along with this software.
 *      If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com.noticemc.noticeconnect.files

import com.typesafe.config.ConfigRenderOptions
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.hocon.Hocon
import kotlinx.serialization.hocon.encodeToConfig
import org.spongepowered.configurate.CommentedConfigurationNode
import org.spongepowered.configurate.hocon.HoconConfigurationLoader
import java.io.BufferedWriter
import java.io.FileOutputStream
import java.io.OutputStreamWriter
import java.io.PrintWriter
import java.nio.file.Path
import kotlin.io.path.createFile
import kotlin.io.path.exists

object ObjectMapperExample {

    fun main(path: Path) {
        val ver = "1.0.0"
        val config = Config(ver,
            Message(listOf("<gray>[<yellow>初見さんいらっしゃい</yellow>] <light_purple><name>さんがはじめて<currentServerName>サーバーに入りました!"),
                listOf("<yellow><gold><name></gold>さんが<gold>サーバー<gold><gray>(<currentServerName>)</gray>にやってきました!"),
                listOf("<yellow><gold><name></gold>さんが<gold>サーバー<gold><gray>(<currentServerName>)</gray>から離れました")),
            Database("NoticeConnect"),
            HashMap(),
            Discord("", "", DiscordMessage("%(PlayerName) さんがサーバーに初参加です", "%(PlayerName) さんがサーバーに参加しました", "%(PlayerName) さんがサーバーから退出しました")))
        val encode = hocon.encodeToConfig(config)

        val renderOptions = ConfigRenderOptions.defaults().setOriginComments(false).setComments(false).setFormatted(true).setJson(false)
        val string = encode.root().render(renderOptions)
        if (!path.exists()) {
            path.createFile()
            val fw = PrintWriter(BufferedWriter(OutputStreamWriter(FileOutputStream(path.toFile()), "UTF-8")))
            fw.write(string)
            fw.close()
        } else {
            val verNode: CommentedConfigurationNode? = HoconConfigurationLoader.builder().path(path).build().load().node("version")
            if (verNode == null || verNode.string != ver) {
                path.toFile().deleteOnExit()
                val fw = PrintWriter(BufferedWriter(OutputStreamWriter(FileOutputStream(path.toFile()), "UTF-8")))
                fw.write(string)
                fw.close()
            }
        }
    }

}

@OptIn(ExperimentalSerializationApi::class)
private val hocon = Hocon

@Serializable
data class DiscordMessage(val firstJoin: String, val join: String, val left: String)

@Serializable
data class Discord(val token: String, val channel: String, val message: DiscordMessage)

@Serializable
data class Message(val firstJoin: List<String>, val join: List<String>, val left: List<String>)

@Serializable
data class Database(val database: String)

@Serializable
data class Config(val version: String, val message: Message, val database: Database, val replace: Map<String, String>, val discord: Discord)